huaweicloud/services/dc/resource_huaweicloud_dc_virtual_interface.go

```diff
@@ -0,0 +1,458 @@
+		EnableNqa:           d.Get("enable_nqa").(bool),
+		LagId:               d.Get("lag_id").(string),
+		ResourceTenantId:    d.Get("resource_tenant_id").(string),
+		EnterpriseProjectId: common.GetEnterpriseProjectID(d, cfg),
+	}
+}
+
+func resourceVirtualInterfaceCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	client, err := cfg.DcV3Client(cfg.GetRegion(d))
+	if err != nil {
+		return diag.Errorf("error creating DC v3 client: %s", err)
+	}
+
+	opts := buildVirtualInterfaceCreateOpts(d, cfg)
+	resp, err := interfaces.Create(client, opts)
+	if err != nil {
+		return diag.Errorf("error creating virtual interface: %s", err)
+	}
+	d.SetId(resp.ID)
+
+	return resourceVirtualInterfaceRead(ctx, d, meta)
+}
+
+func resourceVirtualInterfaceRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	client, err := cfg.DcV3Client(cfg.GetRegion(d))
+	if err != nil {
+		return diag.Errorf("error creating DC v3 client: %s", err)
+	}
+
+	interfaceId := d.Id()
+	resp, err := interfaces.Get(client, interfaceId)
+	if err != nil {
+		return common.CheckDeletedDiag(d, err, "virtual interface")
+	}
+	log.Printf("[Lance] The response of virtual interface is: %#v", resp)
+
+	mErr := multierror.Append(nil,
+		d.Set("region", cfg.GetRegion(d)),
+		d.Set("vgw_id", resp.VgwId),
+		d.Set("type", resp.Type),
+		d.Set("route_mode", resp.RouteMode),
+		d.Set("vlan", resp.Vlan),
+		d.Set("bandwidth", resp.Bandwidth),
+		d.Set("remote_ep_group", resp.RemoteEpGroup),
+		d.Set("name", resp.Name),
+		d.Set("description", resp.Description),
+		d.Set("direct_connect_id", resp.DirectConnectId),
+		d.Set("service_type", resp.ServiceType),
+		d.Set("local_gateway_v4_ip", resp.LocalGatewayV4Ip),
+		d.Set("remote_gateway_v4_ip", resp.RemoteGatewayV4Ip),
+		d.Set("address_family", resp.AddressFamily),
+		d.Set("local_gateway_v6_ip", resp.LocalGatewayV6Ip),
+		d.Set("remote_gateway_v6_ip", resp.RemoteGatewayV6Ip),
+		d.Set("asn", resp.BgpAsn),
+		d.Set("bgp_md5", resp.BgpMd5),
+		d.Set("enable_bfd", resp.EnableBfd),
+		d.Set("enable_nqa", resp.EnableNqa),
+		d.Set("lag_id", resp.LagId),
+		d.Set("resource_tenant_id", resp.TenantId),
+		d.Set("enterprise_project_id", resp.EnterpriseProjectId),
+		d.Set("device_id", resp.DeviceId),
+		d.Set("status", resp.Status),
+		d.Set("created_at", resp.CreatedAt),
+	)
+
+	if err = mErr.ErrorOrNil(); err != nil {
+		return diag.Errorf("error saving virtual interface fields: %s", err)
+	}
+	return nil
+}
+
+func updateVirtualInterface(client *golangsdk.ServiceClient, interfaceId string, opts interfaces.UpdateOpts) error {
+	_, err := interfaces.Update(client, interfaceId, opts)
+	if err != nil {
+		return fmt.Errorf("error updating virtual interface (%s): %s", interfaceId, err)
+	}
+	return nil
+}
+
+func updateVirtualInterfaceBasicInfo(client *golangsdk.ServiceClient, d *schema.ResourceData) error {
```

Comment:
suggest merge updateVirtualInterface and updateVirtualInterfaceBasicInfo to one function.
or change one function which build the params

Reply:
If so, the funtion will hard to read.
I rename the function name: `updateVirtualInterfaceBasicParams`

Reply:
done.